File: week01-what-is-cpp/2-memory-cpp-vs-java/Memory.java

/**
 * Demonstrates that memory usage in Java is surprising.
 */

class Point {
	private int x;
	private int y;
	public Point() { x = y = 0; }	
};


public class Memory {
	static final int SIZE=125*1000*1024; // 125000 KB

	public static void main(String[] args) throws InterruptedException {
		System.out.println("Before new");
		Thread.sleep(3000);

		Point[] p = new Point[SIZE];
		for (int  i=0; i<SIZE; ++i)
			p[i] = new Point();
		
		System.out.println("After new");

		// The free memory should drop by 2 million KB (?)

		Thread.sleep(5000);
	}
}

File: week01-what-is-cpp/2-memory-cpp-vs-java/Memory.cpp
/**
 * Demonstrates that memory usage in C++ is accurate.
 * @author erelsgl
 */

#include <iostream>
#include <thread>
#include <chrono>
using namespace std;

constexpr int SIZE=125*1000*1024; // 125000 KB

class Point {
	private:
		int x;
		int y;
	public:
		Point() { x = y = 0; }	
};

int main() {
	cout << "Before new" << endl;
	this_thread::sleep_for(chrono::seconds(3));

	Point* p = new Point[SIZE];
	for (int  i=0; i<SIZE; ++i)
		p[i] = Point();

	cout << "After new" << endl;
	// The free memory should drop by a million KB.

	this_thread::sleep_for(chrono::seconds(5));
}
